File: Algorithms/singleNumberII.py
class Solution:
    # @param A, a list of integer
    # @return an integer
    def singleNumber(self, A):
        result = 0
        for i in range(32):
            tmp = 0
            bitOps = 1 << i
            for j in A:
                j = j if j > 0 else -j
                if j & bitOps:
                    tmp += 1
            result += (tmp % 3) << i
        flag = sum(1 for i in A if i < 0)
        return result * (-1 if flag % 3 else 1)

File: Algorithms/singleNumberII.cpp
#include "leetcode.h"
class Solution {
public:
    int singleNumber(int A[], int n) {
        int result = 0;
        for (int j=0; j < 32; ++j) {
            int tmp = 0;
            int bitOps = 1 << j;
            for (int i=0; i < n; ++i) {
                if (A[i] & bitOps) {
                    tmp ++;
                }
            }
            result += (tmp%3) << j;
        }
        return result;
    }
};
int main() {
    Solution s;
    int A[] = { 3, 3, 3, 4};
    cout << s.singleNumber(A, 4) << endl;
    int B[] = {-2,-2,1,1,-3,1,-3,-3,-4,-2};
    cout << s.singleNumber(B, 10) << endl;
    int C[] = {43,16,45,89,45,-2147483648,45,2147483646,-2147483647,-2147483648,43,2147483647,-2147483646,-2147483648,89,-2147483646,89,-2147483646,-2147483647,2147483646,-2147483647,16,16,2147483646,43};
    cout << s.singleNumber(C, 25) << endl;

    return 0;
}
